pkg/impersonation/impersonation.go

```diff
@@ -131,44 +132,44 @@ func (i *Impersonator) createNamespace() error {
 	return err
 }
 
-func (i *Impersonator) createRole() (*rbacv1.ClusterRole, error) {
+func (i *Impersonator) checkAndCreateRole() (*rbacv1.ClusterRole, error) {
 	name := ImpersonationPrefix + i.user.GetUID()
-	r, err := i.clusterContext.RBAC.ClusterRoles("").Controller().Lister().Get("", name)
+	rules := []rbacv1.PolicyRule{{
+		Verbs:         []string{"impersonate"},
+		APIGroups:     []string{""},
+		Resources:     []string{"users"},
+		ResourceNames: []string{i.user.GetUID()},
+	}}
+
+	if groups := i.user.GetGroups(); len(groups) > 0 {
+		rules = append(rules, rbacv1.PolicyRule{
+			Verbs:         []string{"impersonate"},
+			APIGroups:     []string{""},
+			Resources:     []string{"groups"},
+			ResourceNames: groups,
+		})
+	}
+	extras := i.user.GetExtra()
+	if principalids, ok := extras["principalid"]; ok {
+		rules = append(rules, rbacv1.PolicyRule{
+			Verbs:         []string{"impersonate"},
+			APIGroups:     []string{"authentication.k8s.io"},
+			Resources:     []string{"userextras/principalid"},
+			ResourceNames: principalids,
+		})
+	}
+	if usernames, ok := extras["username"]; ok {
+		rules = append(rules, rbacv1.PolicyRule{
+			Verbs:         []string{"impersonate"},
+			APIGroups:     []string{"authentication.k8s.io"},
+			Resources:     []string{"userextras/username"},
+			ResourceNames: usernames,
+		})
+	}
+	role, err := i.clusterContext.RBAC.ClusterRoles("").Controller().Lister().Get("", name)
```

Comment:
Technically not part of the change itself, but it looks like `err` is not handled fully here - if `err` is not `IsNotFound` but is not `nil` we proceed as if there was no error.
Suggesting additional error handling to be added.

Reply:
Actually, disregard the "technically not part of the change itself" note. Previous code covered this error handling with `return r, err` at the end of func.

Reply:
fixed in the refactor